src/main/java/com/google/devtools/build/remote/client/RemoteClient.java

```diff
@@ -14,35 +14,64 @@
 
 package com.google.devtools.build.remote.client;
 
+import static java.nio.charset.StandardCharsets.UTF_8;
+
 import com.beust.jcommander.JCommander;
 import com.beust.jcommander.ParameterException;
+import com.google.common.base.CharMatcher;
+import com.google.common.base.Joiner;
+import com.google.common.escape.CharEscaperBuilder;
+import com.google.common.escape.Escaper;
 import com.google.common.hash.Hashing;
 import com.google.devtools.build.remote.client.RemoteClientOptions.CatCommand;
 import com.google.devtools.build.remote.client.RemoteClientOptions.GetDirCommand;
 import com.google.devtools.build.remote.client.RemoteClientOptions.GetOutDirCommand;
 import com.google.devtools.build.remote.client.RemoteClientOptions.LsCommand;
 import com.google.devtools.build.remote.client.RemoteClientOptions.LsOutDirCommand;
+import com.google.devtools.build.remote.client.RemoteClientOptions.ShowActionCommand;
+import com.google.devtools.build.remote.client.RemoteClientOptions.ShowActionResultCommand;
+import com.google.devtools.remoteexecution.v1test.Action;
+import com.google.devtools.remoteexecution.v1test.ActionResult;
+import com.google.devtools.remoteexecution.v1test.Command;
+import com.google.devtools.remoteexecution.v1test.Command.EnvironmentVariable;
 import com.google.devtools.remoteexecution.v1test.Digest;
 import com.google.devtools.remoteexecution.v1test.Directory;
 import com.google.devtools.remoteexecution.v1test.DirectoryNode;
 import com.google.devtools.remoteexecution.v1test.FileNode;
 import com.google.devtools.remoteexecution.v1test.OutputDirectory;
+import com.google.devtools.remoteexecution.v1test.OutputFile;
 import com.google.devtools.remoteexecution.v1test.RequestMetadata;
 import com.google.devtools.remoteexecution.v1test.ToolDetails;
 import com.google.devtools.remoteexecution.v1test.Tree;
+import com.google.protobuf.TextFormat;
+import java.io.FileInputStream;
 import java.io.FileOutputStream;
 import java.io.IOException;
+import java.io.InputStreamReader;
 import java.io.OutputStream;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.stream.Collectors;
 
 /** A standalone client for interacting with remote caches in Bazel. */
 public class RemoteClient {
 
-  AbstractRemoteActionCache cache;
-  DigestUtil digestUtil;
+  private final AbstractRemoteActionCache cache;
+  private final DigestUtil digestUtil;
+
+  private static final Joiner SPACE_JOINER = Joiner.on(' ');
+  private static final Escaper STRONGQUOTE_ESCAPER =
+      new CharEscaperBuilder().addEscape('\'', "'\\''").toEscaper();
+  private static final CharMatcher SAFECHAR_MATCHER =
```

Comment:
Would it make sense to move this into a separate class, the way Bazel's one works?
It is quite likely we'll need Windows support and the way they've set up their abstraction seems reasonable (might also be easier to keep those in sync)

Reply:
I think that's a good idea. I'll write some tests too.

A note on Windows support: While there is a Windows command describer like the Linux one in Bazel
[here](https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/util/CommandFailureUtils.java;l=88?q=CommandFailureUtils.java), escaping the command appears to have been a TODO for a while (2010). Windows will probably also require other command formatting work like changing forward slashes to back slashes.

Reply:
Yes, creating Windows support is not currently part of our mandate - but let's position ourselves to benefit from changes when they do occur in Bazel.

Reply:
Done.

Reply:
Done.